File: src/python/problems/find_difference_test.py
#https://leetcode.com/problems/find-the-difference-of-two-arrays

from core.problem_base import *

class FindDifference(ProblemBase):
    def Solution(self, nums1: list[int], nums2: list[int]) -> list[list[int]]:
        set1 = set(nums1)
        set2 = set(nums2)

        for i in range(len(nums1)):
            if nums1[i] in set2:
                set1.remove(nums1[i])
                set2.remove(nums1[i])
        
        return [list(set1), list(set2)]

if __name__ == '__main__':
    TestGen(FindDifference) \
        .Add(lambda tc: tc.Param([1,2,3]).Param([2,4,6]).Result([[1,3],[4,6]])) \
        .Add(lambda tc: tc.Param([1,2,3,3]).Param([1,1,2,2]).Result([[3],[]])) \
        .Run()

File: src/go/problems/find_difference_test.go
// https://leetcode.com/problems/find-the-difference-of-two-arrays

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type FindDifference struct{}

func TestFindDifference(t *testing.T) {
	gen := core.TestSuite[FindDifference]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]int{1, 2, 3}).Param([]int{2, 4, 6}).Result([][]int{{1, 3}, {4, 6}})
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{1, 2, 3, 3}).Param([]int{1, 1, 2, 2}).Result([][]int{{3}, {}})
	}).Run(t)
}

func (FindDifference) Solution(nums1 []int, nums2 []int) [][]int {
	createHash := func(nums []int) map[int]struct{} {
		set := map[int]struct{}{}
		for _, item := range nums {
			set[item] = struct{}{}
		}
		return set
	}

	filter := func(set1 map[int]struct{}, set2 map[int]struct{}) []int {
		result := []int{}
		for key := range set1 {
			if _, ok := set2[key]; !ok {
				result = append(result, key)
			}
		}
		return result
	}

	set1 := createHash(nums1)
	set2 := createHash(nums2)

	result := make([][]int, 2)
	result[0] = filter(set1, set2)
	result[1] = filter(set2, set1)

	return result
}
